# Introduction

In this notebook, A comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store will be done.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [262]:
import pandas as pd
import plotly.express as px

# Notebook Presentation

In [263]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [264]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows

In [265]:
df_apps.shape    #It shows 10841 rows and 12 cols

(10841, 12)

In [266]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [267]:
df_apps.sample(5)

App          Category  Rating  Reviews  Size_MBs  \
10392                slither.io              GAME    4.40  5234825     36.00   
3604               BIG Launcher   PERSONALIZATION    4.30      881      5.90   
10155  trivago: Hotels & Travel  TRAVEL_AND_LOCAL    4.20   219848     12.00   
6802   Remote CT - Smart Remote             TOOLS    3.50     3988     11.00   
1865     IKORO FH App Rosenheim            FAMILY     NaN       22     29.00   

          Installs  Type  Price Content_Rating           Genres  \
10392  100,000,000  Free      0       Everyone           Action   
3604        10,000  Paid  $9.99       Everyone  Personalization   
10155   50,000,000  Free      0       Everyone   Travel & Local   
6802     1,000,000  Free      0       Everyone            Tools   
1865         1,000  Free      0       Everyone        Education   

            Last_Updated         Android_Ver  
10392  November 14, 2017          2.3 and up  
3604       June 20, 2018          2.1 and up  
10155     August 2, 2018  Varies with device  
6802   December 22, 2015          4.4 and up  
1865    February 6, 2018        2.3.3 and up

### Drop Unused Columns

Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [268]:
df_apps.drop(['Last_Updated','Android_Ver'],axis = 1, inplace = True)
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres'],
      dtype='object')

### Find and Remove NaN values in Ratings

How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [269]:
df_apps.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size_MBs             0
Installs             0
Type                 1
Price                0
Content_Rating       0
Genres               0
dtype: int64

In [270]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [271]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]
print(duplicated_rows.shape)
duplicated_rows.head()

(476, 10)


App Category  Rating  Reviews  \
946                       420 BZ Budeze Delivery  MEDICAL    5.00        2   
1133                                 MouseMingle   DATING    2.70        3   
1196  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL    4.40        8   
1231                                Sway Medical  MEDICAL    5.00        3   
1247              Chat Kids - Chat Room For Kids   DATING    4.70        6   

      Size_MBs Installs  Type   Price Content_Rating   Genres  
946      11.00      100  Free       0     Mature 17+  Medical  
1133      3.90      100  Free       0     Mature 17+   Dating  
1196      6.50      100  Paid  $12.99       Everyone  Medical  
1231     22.00      100  Free       0       Everyone  Medical  
1247      4.90      100  Free       0     Mature 17+   Dating

In [272]:
df_apps_clean[df_apps_clean.App =='Instagram']


App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10809  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

In [273]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App','Type','Price'])
df_apps_clean.shape

(8199, 10)

# Find Highest Rated Apps

Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [274]:
df_apps_clean.head(4)
df_apps_clean.sort_values('Rating', ascending=False).head(5)

App     Category  Rating  Reviews  Size_MBs Installs  \
21    KBA-EZ Health Guide      MEDICAL    5.00        4     25.00        1   
1230         Sway Medical      MEDICAL    5.00        3     22.00      100   
1227    AJ Men's Grooming    LIFESTYLE    5.00        2     22.00      100   
1224       FK Dedinje BGD       SPORTS    5.00       36      2.60      100   
1223      CB VIDEO VISION  PHOTOGRAPHY    5.00       13      2.60      100   

      Type Price Content_Rating       Genres  
21    Free     0       Everyone      Medical  
1230  Free     0       Everyone      Medical  
1227  Free     0       Everyone    Lifestyle  
1224  Free     0       Everyone       Sports  
1223  Free     0       Everyone  Photography

Only apps with very few reviews (and a low number on installs) have perfect 5 star ratings (most likely by friends and family).

# Find 5 Largest Apps in terms of Size (MBs)

 What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [275]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head(5)

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [276]:
df_apps_clean.sort_values('Reviews', ascending=False).head(5)

App       Category  Rating  \
10805                                  Facebook         SOCIAL    4.10   
10785                        WhatsApp Messenger  COMMUNICATION    4.40   
10806                                 Instagram         SOCIAL    4.50   
10784  Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
10650                            Clash of Clans           GAME    4.60   

        Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805  78158306      5.30  1,000,000,000  Free     0           Teen   
10785  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806  66577313      5.30  1,000,000,000  Free     0           Teen   
10784  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650  44891723     98.00    100,000,000  Free     0   Everyone 10+   

              Genres  
10805         Social  
10785  Communication  
10806         Social  
10784  Communication  
10650       Strategy

Reviews shows the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app! 🤔

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [277]:
rating = df_apps_clean.Content_Rating.value_counts()
rating.describe

<bound method NDFrame.describe of Content_Rating
Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: count, dtype: int64>

In [278]:
fig = px.pie(labels=rating.index, values=rating.values, title="Content Rating", names=rating.index)
fig.update_traces(textposition='outside', textinfo='percent+label')

In [279]:
#Donut Chart
fig = px.pie(labels=rating.index, values=rating.values, title="Content Rating", names=rating.index, hole = 0.6)
fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')

# Numeric Type Conversion: Examine the Number of Installs

How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations. 

Convert the number of installations (the Installs column) to a numeric data type.

In [280]:
df_apps_clean.Installs.describe()   #To check datatype of Installs column data

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [281]:
df_apps_clean.info()   #To check datatype of Installs column data

<class 'pandas.core.frame.DataFrame'>
Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


Both of these show that we are dealing with a non-numeric data type. In this case, the type is "object".

If we take two of the columns, say Installs and the App name, we can count the number of entries per level of installations with .groupby() and .count(). However, because we are dealing with a non-numeric data type, the ordering is not helpful. The reason Python is not recognising our installs as numbers is because of the comma (,) characters.


In [282]:
df_apps_clean.loc[:, 'Installs'] = df_apps_clean['Installs'].str.replace(',', "")

In [283]:
df_apps_clean.loc[:, 'Installs'] = pd.to_numeric(df_apps_clean['Installs'])

In [284]:
df_apps_clean[['App','Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [285]:
df_apps_clean.loc[:,'Price'] = df_apps_clean.Price.str.replace('$',"")
df_apps_clean.loc[:,'Price'] = pd.to_numeric(df_apps_clean.Price)


### The most expensive apps sub $250

In [286]:

df_apps_clean = df_apps_clean[df_apps_clean['Price'] <250]
df_apps_clean.sort_values('Price', ascending=False).head(5)

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

     Installs  Type Price Content_Rating     Genres  
2281     1000  Paid 79.99       Everyone    Medical  
1407      100  Paid 39.99       Everyone    Medical  
2629     1000  Paid 37.99       Everyone  Lifestyle  
2481     1000  Paid 33.99       Everyone    Medical  
2463     1000  Paid 29.99       Everyone    Medical

### Highest Grossing Paid Apps (ballpark estimate)

In [287]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type Price Content_Rating                     Genres  \
9220  10000000  Paid  6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid  0.99     Mature 17+                     Action   
7151   1000000  Paid  6.99     Mature 17+                     Action   
7477   1000000  Paid  5.99       Everyone                Photography   
7977   1000000  Paid  5.99       Everyone                  Lifestyle   
6594   1000000  Paid  4.99       Everyone                     Action   
6082    500000  Paid  5.99       Everyone                    Weather   
7954   1000000  Paid  2.99       Everyone                   Strategy   
7633   1000000  Paid  2.99           Teen                     Action   
6746   1000000  Paid  2.99   Everyone 10+    Card;Action & Adventure   

     Revenue_Estimate  
9220    69,900,000.00  
8825     9,900,000.00  
7151     6,990,000.00  
7477     5,990,000.00  
7977     5,990,000.00  
6594     4,990,000.00  
6082     2,995,000.00  
7954     2,990,000.00  
7633     2,990,000.00  
6746     2,990,000.00

The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

In [288]:
df_apps_clean.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Revenue_Estimate'],
      dtype='object')

In [289]:
df_apps_clean.Category.nunique()

33

In [291]:
top_10_category = df_apps_clean.Category.value_counts()[:10]
top_10_category

Category
FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: count, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [293]:
px.bar(x = top_10_category.index, y = top_10_category.values)

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [307]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace = True)
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index, title = "Category Popularity",
               orientation='h')
h_bar.show()


Now we see that Games and Tools are actually the most popular categories. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is.

### Category Concentration - Downloads vs. Competition

* First, create a DataFrame that has the number of apps in one column and the number of installs in another:
* Create scatter plot
* Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [309]:
category_apps = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
merged_df = pd.merge(category_apps, category_installs, on='Category', how="inner")
print(f'The dimensions of the DataFrame are: {merged_df.shape}')
merged_df.sort_values('Installs', ascending=False)[:10]


The dimensions of the DataFrame are: (33, 2)


App     Installs
Category                             
GAME                 910  13858762717
COMMUNICATION        257  11039241530
TOOLS                719   8099724500
PRODUCTIVITY         301   5788070180
SOCIAL               203   5487841475
PHOTOGRAPHY          263   4649143130
FAMILY              1606   4437554490
VIDEO_PLAYERS        148   3916897200
TRAVEL_AND_LOCAL     187   2894859300
NEWS_AND_MAGAZINES   204   2369110650

In [312]:
scatter = px.scatter(merged_df, # data
                    x='App', # column name
                    y='Installs', # column name
                    title='Category Concentration',
                    size='App',
                    hover_name=merged_df.index,
                    color='Installs')

scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)", yaxis_title="Installs", yaxis=dict(type='log'))

scatter.show()

The categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

# Extracting Nested Data from a Column

How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values?

In [319]:
stack = df_apps_clean.Genres.str.split(";", expand = True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')
stack.describe()

We now have a single column with shape: (8564,)
Number of genres: 53


count      8564
unique       53
top       Tools
freq        719
dtype: object

# Colour Scales in Plotly Charts - Competition in Genres

Create this chart with the Series containing the genre data

* Set the colour scale using the color_continuous_scale parameter.
* Make the color axis disappear by using coloraxis_showscale.

In [329]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres.values[:15], # count
             title='Top Genres',
             hover_name=num_genres.index[:15],
             color=num_genres.values[:15],
             color_continuous_scale='Agsunset')

bar.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps', coloraxis_showscale=False)

bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [330]:
df_apps_clean.Type.value_counts()

Type
Free    7595
Paid     589
Name: count, dtype: int64

We see that the majority of apps are free on the Google Play Store. But perhaps some categories have more paid apps than others. Let’s investigate. We can group our data first by Category and then by Type. Then we can add up the number of apps per each type.

In [331]:
df_free_vs_paid = df_apps_clean.groupby(['Category', 'Type'], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.head(5)

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

In [340]:
g_bar = px.bar(df_free_vs_paid,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))

g_bar.show()

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.


# Plotly Box Plots: Lost Downloads for Paid Apps

Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?


In [346]:
box = px.box(df_apps_clean, y='Installs', x='Type', color='Type', notched=True,points='all',
             title='How Many Downloads are Paid Apps Giving Up?')
box.update_layout(yaxis=dict(type='log'))
box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower. Should we give on making Paid app. Let’s see how much revenue we would estimate per category.

# Plotly Box Plots: Revenue by App Category

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?


In [349]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))
box.show()

# How Much Can one Charge? Examine Paid App Pricing Strategies by Category

What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps.

In [350]:
df_paid_apps.Price.median()

np.float64(2.99)

In [351]:
box = px.box(df_paid_apps,
             x='Category',
             y="Price",
             title='Price per Category')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))

box.show()

The median price for an Android app is $2.99. However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of $5.49. In contrast, Personalisation apps are quite cheap on average at $1.49. Other categories which higher median prices are Business ($4.99) and Dating ($6.99). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.